### PR TITLE
Case 20866: gltf color attribute

### PR DESCRIPTION
**Pull request by raveenajain**
https://highfidelity.manuscript.com/f/cases/20866/glTF-color-attribute
https://github.com/highfidelity/hifi_tests/pull/336

### CODE REVIEW

**Comment by raveenajain** [comment]
Test Plan: 

1 - Run interface. Go to Edit -> Running Scripts -> From URL
2 - add https://raw.githubusercontent.com/raveenajain/hifi_tests/gltf_tests/tests/content/entity/model/modelReaders/gltfReader/gltfTestSuite/boxes/testBoxColors.js 
3 - output seen above should match https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/BoxVertexColors/screenshot/screenshot.png
OR (if you want to compare 3D models)
i - download https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/BoxVertexColors/glTF
ii - double click to open BoxVertexColors.gltf (i.e if on Windows will open with 3D Viewer)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16807/HighFidelity-Beta-PR14786-e112bf1.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16807/HighFidelity-Beta-Interface-PR14786-e112bf1.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16807/HighFidelity-Beta-PR14786-e112bf1.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16807/HighFidelity-Beta-Interface-PR14786-e112bf1.exe

**Comment by raveenajain** [comment]
Jenkins, rebuild this please.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16814/HighFidelity-Beta-PR14786-e112bf1.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16814/HighFidelity-Beta-PR14786-e112bf1.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16814/HighFidelity-Beta-Interface-PR14786-e112bf1.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16814/HighFidelity-Beta-PR14786-e112bf1.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16814/HighFidelity-Beta-Interface-PR14786-e112bf1.exe

File: libraries/fbx/src/GLTFSerializer.cpp [hunk 1]
```diff
@@ -831,6 +831,27 @@ bool GLTFSerializer::buildGeometry(HFMModel& hfmModel, const QUrl& url) {
                         for (int n = 0; n < normals.size(); n = n + 3) {
                             mesh.normals.push_back(glm::vec3(normals[n], normals[n + 1], normals[n + 2]));
                         }
+                    } else if (key == "COLOR_0") {
+                        QVector<float> colors;
```
Comment (AndrewMeadows) [inline libraries/fbx/src/GLTFSerializer.cpp]: Not a blocker: I know we're already using QVector<> all over this file, but for **future work** consider switching over to `std::vector<>`.  There are a few good reasons to not use Qt containers when possible.  See [this article](https://www.cleanqt.io/blog/exploring-qt-containers) for some of those reasons or maybe watch the CPP-Con talk by Giuseppe D'Angelo called ["Effective Qt: 2017 Edition"](https://www.youtube.com/watch?v=uZ68dX1-sVc).

File: libraries/fbx/src/GLTFSerializer.cpp [hunk 2]
```diff
@@ -831,6 +831,27 @@ bool GLTFSerializer::buildGeometry(HFMModel& hfmModel, const QUrl& url) {
...
+                            bufferview.byteOffset + accBoffset, 
+                            accessor.count, 
+                            colors,
+                            accessor.type, 
+                            accessor.componentType);
+                        if (!success) {
+                            qWarning(modelformat) << "There was a problem reading glTF COLOR_0 data for model " << _url;
+                            continue;
+                        }
+                        if (accessor.type == 3) {
```
Comment (AndrewMeadows) [inline libraries/fbx/src/GLTFSerializer.cpp]: Don't use a magic number `3` here.  Instead do:  `if (accessor.type == GLTFAccessorType::VEC4) {`.

File: libraries/fbx/src/GLTFSerializer.cpp [hunk 3]
```diff
@@ -831,6 +831,27 @@ bool GLTFSerializer::buildGeometry(HFMModel& hfmModel, const QUrl& url) {
...
+                            for (int n = 0; n < colors.size(); n = n + 4) {
+                                mesh.colors.push_back(glm::vec3(colors[n], colors[n + 1], colors[n + 2]));
+                            }
+                        } else {
+                            for (int n = 0; n < colors.size(); n = n + 3) {
+                                mesh.colors.push_back(glm::vec3(colors[n], colors[n + 1], colors[n + 2]));
+                            }
+                        }
```
Comment (AndrewMeadows) [inline libraries/fbx/src/GLTFSerializer.cpp]: These two different for loops could be condensed into one like so:
```
int stride = (accessor.type == GLTFAccessorType::VEC4) ? 4 : 3;
for (int n = 0; n < colors.size() - 3; n += stride) {
    mesh.colors.push_back(glm::vec3(colors[n], colors[n + 1], colors[n + 2]));  
}
```
Note, the comparison check in the for loop requires `colors.size() - 3` else you risk stepping out of bounds when you do `color[n + 2]`.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16968/HighFidelity-Beta-PR14786-997660d.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16968/HighFidelity-Beta-PR14786-997660d.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16968/HighFidelity-Beta-Interface-PR14786-997660d.exe

**Pull request merged**